i#1312 AVX-512 support: Add remaining v(p)mov and vpshuf* opcodes.

hgreving2304 opened the pull request:
This patch contains a minor compatibility change that now splits the existing AVX general register
versions of vmovd/vmovq into the opcodes vmovd/vmovq.

Adds the evex-promoted opcodes vmovd, vmovq, vpshufhw, vpshufd, vpshuflw,
vpshufb, vpmovsxbw, vpmovsxbd, vpmovsxbq, vpmovsxwd, vpmovsxwq, vpmovsxdq,
vpmovzxbw, vpmovzxbd, vpmovzxbq, vpmovzxwd, vpmovzxwq, and vpmovzxdq.

Adds the new AVX-512 opcodes vpmovqb, vpmovsqb, vpmovusqb, vpmovqw, vpmovsqw,
vpmovusqw, vpmovqd, vpmovsqd, vpmovusqd, vpmovdb, vpmovsdb, vpmovusdb, vpmovdw,
vpmovsdw, vpmovusdw, vpmovwb, vpmovswb, vpmovuswb, vpmovm2b, vpmovm2w, vpmovm2d,
vpmovm2q, vpmovb2m, vpmovw2m, vpmovd2m, and vpmovq2m.

Adds new letter codes 'i' for quarter and 'j' for eighth.

Along with adding the AVX-512 evex-promoted opcodes, this patch fixes incorrect AVX
constraints of vpmovsxbw, vpmovsxbd, vpmovsxbq, vpmovsxwd, vpmovsxwq, vpmovsxdq,
vpmovzxbw, vpmovzxbd, vpmovzxbq, vpmovzxwd, vpmovzxwq, and vpmovzxdq.

Adds the internal opsizes OPSZ_quarter_16_vex32, OPSZ_quarter_16_vex32_evex64,
OPSZ_eighth_16_vex32, and OPSZ_eighth_16_vex32_evex64.

Adds tests for above.

Opcodes have been checked against llvm-mc, binutils/gas/objdump and capstone.

This patch does not include the AVX-512 vshuf* opcodes, which will be included in a
future patch.

Issue: #1312

hgreving2304 commented:
cleancall test failure TBD.
I assumed that this was a local failure specific to Ubuntu 18.

derekbruening commented:
Just looking at the description so far, but it sounds like a lot of different things combined into one PR.  Maybe the two "Fixes..." should be separated out?

derekbruening commented on core/arch/x86/decode_table.c:
```diff
@@ -5356,6 +5384,362 @@ const instr_info_t prefix_extensions[][12] = {
+    {INVALID,      0xf2383218, "(bad)",   xx, xx, xx, xx, xx, no, x, NA},
+    {INVALID,        0x383218, "(bad)",   xx, xx, xx, xx, xx, no, x, NA},
+    {OP_vpmovqb,   0xf3383218, "vpmovqb", Wj_e, xx, KEb, Ve, xx, mrm|evex, x, END_LIST},
+    /* XXX i#3639, tools tend to give source/destination different register size mnemonic.
```
grammar + style nit: s/,/:/
Many times below as well

hgreving2304 replied:
Fixed.

derekbruening commented on core/arch/x86/instr_create.h:
```diff
@@ -1751,6 +1769,28 @@
+ * \param dc The void * dcontext used to allocate memory for the instr_t.
+ * \param d The opnd_t explicit destination operand for the instruction.
+ * \param k The opnd_t explicit mask source operand for the instruction.
+ * \param i The opnd_t explicit second source operand for the instruction, which
```
sentence just ends in "which"

derekbruening commented on core/arch/x86/instr_create.h:
```diff
@@ -1751,6 +1769,28 @@
+ * \param k The opnd_t explicit mask source operand for the instruction.
+ * \param i The opnd_t explicit second source operand for the instruction, which
+ * \param s The opnd_t explicit source operand for the instruction.
+ * must be an immediate integer (opnd_create_immed_int()).
```
dangling sentence end -- probably goes with the "which" above

hgreving2304 replied:
Yes, see above, fixed.

derekbruening commented on core/arch/x86/instr_create.h:
```diff
@@ -1751,6 +1769,28 @@
+ * \param k The opnd_t explicit mask source operand for the instruction.
+ * \param i The opnd_t explicit second source operand for the instruction, which
+ * \param s The opnd_t explicit source operand for the instruction.
+ * must be an immediate integer (opnd_create_immed_int()).
```
I see this same problem in already-committed code for `/** @name 1 explicit destination, 1 mask, 3 sources where the final is an immediate */` -- please fix.

hgreving2304 replied:
Done.

hgreving2304 commented:
> Just looking at the description so far, but it sounds like a lot of different things combined into one PR. Maybe the two "Fixes..." should be separated out?

I have another _big_ patch with more instructions based on this one, the less re-factoring the faster we get those in. The fixes have been done along with the new evex-promoted instructions of the corresponding opcodes. Re-factoring only if you feel strongly about it, I'd prefer to finish the decoding work rather sooner than later :) I do think they could be considered to be in the same context. WDYT

derekbruening commented on core/arch/x86/decode_table.c:
```diff
@@ -3679,16 +3709,15 @@ const instr_info_t prefix_extensions[][12] = {
     {OP_movd,   0x0f7e10, "movd", Ed_q, xx, Pd_q, xx, xx, mrm, x, tpe[51][2]},
     /* movq zeroes the top bits when the destination is an mmx or xmm reg */
     {OP_movq, 0xf30f7e10, "movq", Vdq, xx, Wq_dq, xx, xx, mrm, x, tpe[61][2]},
-    {OP_movd, 0x660f7e10, "movd", Ed_q, xx, Vd_dq, xx, xx, mrm, x, END_LIST},
```
Not sure about this: isn't this 4 bytes w/o rex.w, and 8 bytes w/ rex.w?  I.e., the size looks correct at first glance to me (though its name `d_q` seems to violate the conventions: shouldn't be an underscore...separate issue).

As mentioned earlier, it seems best to separate things like this out from adding the new opcodes.

hgreving2304 replied:
Ok, looks like that we're calling both the 66 0F 7E /r and the  66 REX.W 0F 7E /r version "movd", while Intel calls the REX.W version "movq". That must have confused me, but you're right it should be Ed_q - for now. I've reverted this. Thanks! File an issue to change into separate entry + test?

derekbruening replied:
There are numerous cases where Intel names an opcode differently but we don't bother to split our table entries (see "Alternative opcode name" in opcode.h) so I'm not sure we need to do this.  For first-class disassembly support we might, but then we'd also want to go split all those others (cdge, cqo, jrcxz, cmpxchg16b, etc.).  We already have #1307 for them so this movd/movq would fall under that issue.

hgreving2304 replied:
Ok, see comment below.

hgreving2304 commented:
- Something is off - drdecode results made more sense after the "Ed" change for mod/movq that I just reverted. Sorting this out now.

hgreving2304 commented:
> * Something is off - drdecode results made more sense after the "Ed" change for mod/movq that I just reverted. Sorting this out now.

I see what happened. In order to make it more straight forward, I am no longer combining the 32-bit and the rex.W AVX vmovd into one opcode. Instead, it is now split into vmovd and vmovq. This is not really a fix, but makes it more straight forward when adding vmovd/vmovq for AVX-512. I changed the commit comment and explained it.

hgreving2304 commented:
Done - PTAL. The remaining AVX changes are not separate fixes but really consistent changes along with the evex-promoted instruction that was added (vmovd/vmovq) as well as constraints of the mov instructions.

For the AVX vmovd/vmovq related change, does this need a release note?

derekbruening commented on core/arch/decode.h:
```diff
@@ -398,12 +398,30 @@ enum {
     OPSZ_half_16_vex32,      /* half of 128 bits (XMM or memory);
```
> The AVX general register vmovd/vmovq is not split into the the opcodes vmovd/vmovq.

s/the the/the/

s/not/now/

hgreving2304 replied:
Fixed.

derekbruening commented:
> The remaining AVX changes are not separate fixes but really consistent changes along with the evex-promoted instruction that was added (vmovd/vmovq) as well as constraints of the mov instructions.

Maybe that should be separated out from the others, then, or the evex-promoted split from the new: just seems like a lot to have changes in existing things, including non-AVX512 things, combined with a bunch of new opcodes in the same PR.

derekbruening commented:
> For the AVX vmovd/vmovq related change, does this need a release note?

I would say yes.

hgreving2304 commented:
> > For the AVX vmovd/vmovq related change, does this need a release note?
> 
> I would say yes.

Done.

derekbruening commented on api/docs/release.dox:
```diff
@@ -169,6 +169,8 @@ compatibility changes:
    collect more stats.
  - Moves the immediate operand of the VEX encoded version of vextractps and vinsertps
    to be the first source.
+ - The GPR versions of the vmovd and vmovq AVX opcodes are no longer combined into a
```
I would still vote for splitting out in some fashion the compat changes.

hgreving2304 replied:
I moved the comment in the commit message "This patch contains a minor compatibility change that now splits the existing AVX general register versions of vmovd/vmovq into the opcodes vmovd/vmovq." to the top.
